### PR TITLE
twister: pytest: Parse report file to get testcases

### PR DESCRIPTION
**Pull request by gchwier**
Extended parsing of report.xml file - produced by pytest-twister-harness plugin. Now testcases are properly extracted and added to twister report. Added unit tests.

To test just run:
`./scripts/twister -vv -T samples/subsys/testsuite/pytest/shell -p native_posix -p qemu_x86`
in test reports: `twister.json` and `twiseter_report.xml` you should find testcases: `test_shell_print_help` and `test_shell_print_version`.

Some changes also added to correct error handling. In couple of places I added annotations (just to help my IDE understanding the code).

### CODE REVIEW

File: scripts/pylib/twister/twisterlib/harness.py [hunk 1]
```diff
@@ -397,16 +368,54 @@ def _output_reader(proc):
...
+                self.state = 'error'
+            elif elem_ts.get('skipped') == elem_ts.get('tests'):
+                self.state = 'skipped'
+            else:
+                self.state = 'passed'
+            self.instance.execution_time = float(elem_ts.get('time'))
+
+            for elem_tc in elem_ts.findall('testcase'):
+                tc = self.instance.get_case_or_create(
+                    '{}.{}'.format(self.id, elem_tc.get('name'))
```
Comment (gopiotr) [inline scripts/pylib/twister/twisterlib/harness.py:404]: Please use `f-string` instead of `.format`

File: scripts/pylib/twister/twisterlib/testinstance.py [hunk 2]
```diff
@@ -51,7 +59,7 @@ def __init__(self, testsuite, platform, outdir):
         self.domains = None
 
         self.run = False
-        self.testcases = []
+        self.testcases: list[TestCase] = []
```
Comment (gopiotr) [inline scripts/pylib/twister/twisterlib/testinstance.py:62]: If you use `list[...]`, then probably you should add `from __future__ import annotations` at the beginning for Python 3.8. But on the other hand CI passed, so maybe it is not so necessary...

File: scripts/pylib/twister/twisterlib/harness.py [hunk 3]
```diff
@@ -318,53 +316,26 @@ def _generate_parameters_for_hardware(self, handler):
...
+        ) as proc:
             try:
                 reader_t = threading.Thread(target=self._output_reader, args=(proc,), daemon=True)
                 reader_t.start()
                 reader_t.join(timeout)
                 if reader_t.is_alive():
                     terminate_process(proc)
-                    logger.warning('Timeout has occurred.')
+                    logger.warning('Timeout has occurred. Can be extended in testspec file. '
+                                   f'Currently set to {timeout} seconds.')
```
Comment (gopiotr) [inline scripts/pylib/twister/twisterlib/harness.py:332]: Please use `%` instead of `f-string` in logging message.

File: scripts/pylib/twister/twisterlib/harness.py [hunk 4]
```diff
@@ -397,16 +368,54 @@ def _output_reader(proc):
...
-            self.instance.reason = "Pytest timeout"
-        if self.instance.status in ["error", "failed"]:
-            self.instance.add_missing_case_status("blocked", self.instance.reason)
+    def update_test_status(self):
+        if not self.state:
+            self.instance.testcases = []
+            try:
+                self._parse_report_file(self.report_file)
+            except Exception as e:
+                logger.error(f'Error when parsing file {self.report_file}: {e}')
```
Comment (gopiotr) [inline scripts/pylib/twister/twisterlib/harness.py:377]: Please use `%` instead of `f-string` in logging message.

**Review by gopiotr — DISMISSED** [review]
LGTM. Some minor suggestions

File: scripts/pylib/twister/twisterlib/testinstance.py (hunk 2 above)
Comment (gchwier) [reply scripts/pylib/twister/twisterlib/testinstance.py:62]: I will add this import. For python3.8 it is probable not required (for that case, AFAIK we used that for a conditional annotations, e.g. `self.testcases: list[TestCase] | None = None`)

File: scripts/pylib/twister/twisterlib/harness.py (hunk 3 above)
Comment (gchwier) [reply scripts/pylib/twister/twisterlib/harness.py:332]: For debug level logging it is better to use `%` instead of `f-string`, because of performance. But when using info, warning or higher level of logging, and there is not too much of such a logs it shouldn't be a problem

**Comment by gchwier** [comment]
@gopiotr thanks for review. I updated and rebased PR.

File: scripts/pylib/twister/twisterlib/harness.py (hunk 4 above)
Comment (gchwier) [reply scripts/pylib/twister/twisterlib/harness.py:377]: ditto

File: scripts/pylib/twister/twisterlib/handlers.py [hunk 5]
```diff
@@ -40,6 +40,7 @@
 logger.setLevel(logging.DEBUG)
 
 SUPPORTED_SIMS = ["mdb-nsim", "nsim", "renode", "qemu", "tsim", "armfvp", "xt-sim", "native"]
+SUPPORTED_SIMS_IN_PYTEST = ['native', 'qemu']
```
Comment (hakehuang) [inline scripts/pylib/twister/twisterlib/handlers.py:43]: is this change related to the PR?

File: scripts/pylib/twister/twisterlib/harness.py [hunk 6]
```diff
@@ -63,7 +62,7 @@ def __init__(self):
         self.run_id = None
         self.matched_run_id = False
         self.run_id_exists = False
-        self.instance = None
+        self.instance: TestInstance | None = None
```
Comment (hakehuang) [inline scripts/pylib/twister/twisterlib/harness.py:65]: why add this decoration only?  this will mess the coding style.

File: scripts/pylib/twister/twisterlib/harness.py [hunk 7]
```diff
@@ -234,16 +233,15 @@ def configure(self, instance):
     def pytest_run(self, timeout):
         try:
             cmd = self.generate_command()
-            if not cmd:
```
Comment (hakehuang) [inline scripts/pylib/twister/twisterlib/harness.py:237]: why remove this check?

File: scripts/pylib/twister/twisterlib/harness.py [hunk 8]
```diff
@@ -397,16 +368,52 @@ def _output_reader(proc):
...
-        if self.state:
-            self.instance.status = self.state
-            if self.state in ["error", "failed"]:
-                self.instance.reason = "Pytest failed"
-        else:
-            self.instance.status = "failed"
-            self.instance.reason = "Pytest timeout"
-        if self.instance.status in ["error", "failed"]:
-            self.instance.add_missing_case_status("blocked", self.instance.reason)
+    def update_test_status(self):
```
Comment (hakehuang) [inline scripts/pylib/twister/twisterlib/harness.py:371]:
```suggestion
    def _update_test_status(self):
```

File: scripts/pylib/twister/twisterlib/testinstance.py [hunk 9]
```diff
@@ -3,19 +3,27 @@
 # Copyright (c) 2018-2022 Intel Corporation
 # Copyright 2022 NXP
 # SPDX-License-Identifier: Apache-2.0
-
+from __future__ import annotations
```
Comment (hakehuang) [inline scripts/pylib/twister/twisterlib/testinstance.py:6]: suggest to change the annotation changes to another PR

File: scripts/pylib/twister/twisterlib/harness.py (hunk 7 above)
Comment (gchwier) [reply scripts/pylib/twister/twisterlib/harness.py:237]: in current implementation `cmd` will never be empty. If something went wrong in `self.generate_command()`, then `PytestHarnessException` will be raised and handled in the current method

File: scripts/pylib/twister/twisterlib/handlers.py (hunk 5 above)
Comment (gchwier) [reply scripts/pylib/twister/twisterlib/handlers.py:43]: I moved it to avoid circular importing. It is in this PR, because when added annotations (I know, that annotations could be in separate PR), circular importing happens

File: scripts/pylib/twister/twisterlib/harness.py (hunk 6 above)
Comment (gchwier) [reply scripts/pylib/twister/twisterlib/harness.py:65]: This is annotation to let user know, what kind of object is added to that class. I know, that we are not using annotations in twister, and I do not want to introduce any strong type checking, but rather I wanted to add annotation in couple of places to make that code more readable (cleaner, self-documented, improve code completion in IDE).

File: scripts/pylib/twister/twisterlib/testinstance.py (hunk 9 above)
Comment (gchwier) [reply scripts/pylib/twister/twisterlib/testinstance.py:6]: Here I wanted only to add annotations to improve code completion and readability in places, I changed. Just to do a small step, do not add annotations in the rest of the code. If using annotations in that PR is a blocker, I will remove it and add it in separate PR (or not, if there is no plan to use annotation in twister to not mess the coding style)

**Review by gopiotr — APPROVED** [review]

File: scripts/pylib/twister/twisterlib/testinstance.py (hunk 9 above)
Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/testinstance.py:6]: @gchwier add annotation is good, but maybe in a more consistent way. @PerMac  what do you think?

Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/testinstance.py:6]: @PerMac I mean do we need to create a general task to add anotation to all twisters apis?

**Review by hakehuang — APPROVED** [review]
some thoughts on annotation introducing, not a blocker here

File: scripts/pylib/twister/twisterlib/testinstance.py (hunk 9 above)
Comment (nashif) [reply scripts/pylib/twister/twisterlib/testinstance.py:6]: > I heard several times an advice from seniors that it is good to do such "refactoring" iteratively.

This should be set as a goal first and there needs to be an agreement that we are going this way in general, it should not be left up to individual preferences and choices. 

> Refactoring everything in one go can be a long and potentially error prone proces.

And doing this in an unmanaged way is confusing and makes code very difficult to read and maintain.

> I also saw similar approach already done for twister [#58046 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/58046#discussion_r1202238875) It wasn't annotation related but in general few changes resolving IDE complains.

Not sure how this is related. This seems to be a bug fix in this case.

I am fine with this overall, but we should avoid having different styles in the code and we need to agree on the goals and figure out a roadmap to achieve them.

**Review by nashif — APPROVED** [review]
